File: apis/webarchive.py
#!/usr/bin/env python3
import sys
import json
import requests
import re
import time
from urllib.parse import quote

def fetch_webarchive(domain):
    """Fetch subdomains from Web Archive with improved error handling"""
    subdomains = set()
    
    endpoints = [
        "http://web.archive.org/cdx/search/cdx",
        "https://web.archive.org/cdx/search/cdx"
    ]
    
    params = {
        'url': f'*.{domain}/*',
        'output': 'json',
        'collapse': 'urlkey',
        'limit': 5000,
        'fl': 'original'
    }
    
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': 'application/json',
        'Accept-Language': 'en-US,en;q=0.9'
    }
    
    for endpoint in endpoints:
        try:
            response = requests.get(
                endpoint,
                params=params,
                headers=headers,
                timeout=45,
                verify=False  
            )
            
            if response.status_code == 200:
                try:
                    data = response.json()
                    if data and len(data) > 1:
                        for row in data[1:]:
                            if row and len(row) > 0:
                                url = row[0].lower()
                                # Extract domain from URL
                                pattern = r'(?:https?://)?([a-zA-Z0-9._-]+\.' + re.escape(domain) + ')'
                                matches = re.findall(pattern, url)
                                for match in matches:
                                    if '*' not in match and not match.startswith('www.archive.org'):
                                        subdomains.add(match)
                except json.JSONDecodeError:
                
                    lines = response.text.strip().split('\n')
                    for line in lines[1:]:  # Skip header
                        parts = line.split()
                        if len(parts) > 2:
                            url = parts[2].lower()
                            pattern = r'(?:https?://)?([a-zA-Z0-9._-]+\.' + re.escape(domain) + ')'
                            matches = re.findall(pattern, url)
                            for match in matches:
                                if '*' not in match:
                                    subdomains.add(match)
                
                if subdomains:
                    break
                    
        except (requests.exceptions.RequestException, requests.exceptions.Timeout):
            continue
    
    if not subdomains:
        try:
            timemap_url = f"https://web.archive.org/web/timemap/json?url={domain}&matchType=domain&output=json"
            response = requests.get(timemap_url, timeout=30, headers=headers)
            
            if response.status_code == 200:
                try:
                    data = response.json()
                    for entry in data[1:]:  # Skip header
                        if len(entry) > 2:
                            url = entry[2].lower()
                            pattern = r'([a-zA-Z0-9._-]+\.' + re.escape(domain) + ')'
                            matches = re.findall(pattern, url)
                            for match in matches:
                                subdomains.add(match)
                except:
                    pass
        except:
            pass
    
    return sorted(subdomains)

if __name__ == '__main__':
    if len(sys.argv) != 2:
        sys.exit(0)
    
    domain = sys.argv[1].lower()
    try:
        for subdomain in fetch_webarchive(domain):
            print(subdomain)
    except:
        sys.exit(0)

File: internal/sources/webarchive.go
package sources

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

// WebArchive queries the Wayback Machine CDX API.
type WebArchive struct{}

func (w *WebArchive) Name() string { return "webarchive" }

func (w *WebArchive) Enumerate(ctx context.Context, domain string, client *http.Client) Result {
	start := time.Now()
	domains, attempts, err := retry(3, 3*time.Second, func() ([]string, error) {
		return w.fetch(ctx, domain, client)
	})
	return Result{
		Name:     w.Name(),
		Domains:  dedup(domains),
		Duration: time.Since(start),
		Attempts: attempts,
		Err:      err,
	}
}

var hostRegex = regexp.MustCompile(`(?i)(?:https?://)?([a-zA-Z0-9._-]+)`)

func (w *WebArchive) fetch(ctx context.Context, domain string, client *http.Client) ([]string, error) {
	seen := make(map[string]bool)
	suffix := "." + domain

	endpoints := []string{
		"https://web.archive.org/cdx/search/cdx",
		"http://web.archive.org/cdx/search/cdx",
	}

	params := url.Values{
		"url":      {"*." + domain + "/*"},
		"output":   {"json"},
		"collapse": {"urlkey"},
		"limit":    {"5000"},
		"fl":       {"original"},
	}

	extractFromURLs := func(rawURLs []string) {
		re := regexp.MustCompile(`(?i)(?:https?://)?([a-zA-Z0-9._-]+\.` + regexp.QuoteMeta(domain) + `)`)
		for _, rawURL := range rawURLs {
			m := re.FindStringSubmatch(rawURL)
			if len(m) > 1 {
				h := strings.ToLower(m[1])
				if strings.HasSuffix(h, suffix) || h == domain {
					seen[h] = true
				}
			}
		}
	}

	for _, endpoint := range endpoints {
		reqURL := fmt.Sprintf("%s?%s", endpoint, params.Encode())
		req, err := http.NewRequestWithContext(ctx, "GET", reqURL, nil)
		if err != nil {
			continue
		}
		req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
		req.Header.Set("Accept", "application/json")

		resp, err := client.Do(req)
		if err != nil {
			continue
		}

		if resp.StatusCode == 200 {
			// CDX JSON: [[header], [row1], [row2], ...]
			var rows [][]string
			decodeErr := json.NewDecoder(resp.Body).Decode(&rows)
			resp.Body.Close()

			if decodeErr == nil && len(rows) > 1 {
				rawURLs := make([]string, 0, len(rows)-1)
				for _, row := range rows[1:] {
					if len(row) > 0 {
						rawURLs = append(rawURLs, row[0])
					}
				}
				extractFromURLs(rawURLs)
			}
		} else {
			resp.Body.Close()
		}

		if len(seen) > 0 {
			break
		}
	}

	// Timemap fallback for when CDX returns nothing
	if len(seen) == 0 {
		timemapURL := fmt.Sprintf("https://web.archive.org/web/timemap/json?url=%s&matchType=domain&output=json", domain)
		req, err := http.NewRequestWithContext(ctx, "GET", timemapURL, nil)
		if err == nil {
			req.Header.Set("User-Agent", "Mozilla/5.0")
			resp, err := client.Do(req)
			if err == nil && resp.StatusCode == 200 {
				var data [][]string
				if json.NewDecoder(resp.Body).Decode(&data) == nil {
					re := regexp.MustCompile(`([a-zA-Z0-9._-]+\.` + regexp.QuoteMeta(domain) + `)`)
					for _, row := range data[1:] {
						if len(row) > 2 {
							m := re.FindStringSubmatch(row[2])
							if len(m) > 1 {
								seen[strings.ToLower(m[1])] = true
							}
						}
					}
				}
				resp.Body.Close()
			}
		}
	}

	result := make([]string, 0, len(seen))
	for d := range seen {
		result = append(result, d)
	}
	return result, nil
}
